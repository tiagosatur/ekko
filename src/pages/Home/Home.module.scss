@import "../../styles//theme";

.container {
  padding: 48px 0;
}

.charactersListItem {
  margin: calc(#{$spacer} * 2);
  width: 100%;

  @media (min-width: 768px) {
    width: 47%;
  }
}
