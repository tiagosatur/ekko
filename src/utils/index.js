export { default as ROUTES } from "./routes";
export { default as uniqueId } from "./uniqueId";
export { default as useDate } from "./useDate";
export { default as updateFavorite } from "./updateFavorite";
export { default as populateFavoriteForm } from "./populateFavoriteForm";

export * from "./hooks";
export * from "./manageLocalStorage";
