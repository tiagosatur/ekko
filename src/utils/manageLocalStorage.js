export const loadState = (localStorageName) => {
  const serializedState = localStorage.getItem(localStorageName);
  return serializedState ? { favorites: JSON.parse(serializedState) } : {};
};

export const saveState = (state, localStorageName = "appState") => {
  try {
    const serializedState = JSON.stringify(state.favorites);
    localStorage.setItem(localStorageName, serializedState);
  } catch (e) {
    console.log("sessionStorage error", e);
  }
};

export const manageLocalStorage = (localStorageName) => {
  return {
    saveState: (state) => saveState(state, localStorageName),
    loadState: () => loadState(localStorageName),
  };
};
