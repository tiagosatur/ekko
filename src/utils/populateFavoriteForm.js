import { useDate } from ".";

export default function populateFavoriteForm(favoriteCharacter, setValue) {
  favoriteCharacter &&
    Object.entries(favoriteCharacter).map(([key, val]) => {
      if (typeof val === "object" && val !== null && !Array.isArray(val)) {
        Object.entries(val).map(([subKey, subVal]) => {
          const camelCaseSubKey = `${key}${
            subKey.charAt(0).toUpperCase() + subKey.slice(1)
          } `.replace(/\s/g, "");

          return setValue(camelCaseSubKey, subVal);
        });
      }

      if (key === "created") return setValue(key, useDate.format.iso.date(val));
      return setValue(key, val);
    });
}
