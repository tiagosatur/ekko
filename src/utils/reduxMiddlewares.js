import TYPES from "../store/types";
import mutations from "../store/mutations";

const reduxMiddlewares = {
  addFavoriteMiddlewareInterceptor: ({ getState, dispatch }) => (
    next
  ) => async (action) => {
    const result = next(action);
    if (
      action.type === TYPES.ADD_FAVORITE_SUCCESS ||
      action.type === TYPES.DELETE_FAVORITE_SUCCESS
    ) {
      dispatch(mutations[TYPES.TOGGLE_FAVORITE](action.payload));
    }
    return Promise.resolve(result);
  },
};

export default reduxMiddlewares;
