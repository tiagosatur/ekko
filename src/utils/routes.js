export default {
  HOME: {
    PATH: "/",
    LABEL: "Home",
  },
  CHARACTER: {
    PATH: "/character",
    LABEL: "Character",
  },
  FAVORITE: {
    VIEW: {
      PATH: "/favorites",
      LABEL: "Favorites",
    },
    NEW: {
      PATH: "/favorite/new",
      LABEL: "New favorite",
    },
    UPDATE: {
      PATH: "/favorite/update",
      LABEL: "Update favorite",
    },
  },
};
