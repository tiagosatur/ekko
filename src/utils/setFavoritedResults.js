import store from "../store";

export default function setFavoritedResults(arr) {
  return arr?.map((character) => {
    const match = store
      .getState()
      .favorites.characters.some((favorite) => favorite.id === character.id);
    return {
      ...character,
      isFavorite: match,
    };
  });
}
