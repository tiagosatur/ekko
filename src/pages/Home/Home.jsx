import React, { useEffect } from "react";
import { Row, Col } from "react-flexbox-grid";
import { Link } from "react-router-dom";
import { isEmpty } from "lodash";

import { useAction, useStore, uniqueId, testId, ROUTES } from "../../utils";
import { CharacterCard, Spinner } from "../../components";
import styles from "./Home.module.scss";

export default function Home() {
  const { actions } = useAction();
  const {
    state: {
      characters: { list, charactersLoading },
    },
  } = useStore();

  useEffect(() => {
    isEmpty(list) && actions.getAllcharacters();
  });

  return (
    <div className={styles.container} data-testid={testId.home.container}>
      <Row>
        <Col xs={12}>
          {charactersLoading ? (
            <Spinner />
          ) : (
            <Row>
              <Col xs={12}>
                <ul className={`dFlex flexWrap`}>
                  {list.length > 0 &&
                    list.map((item) => (
                      <li
                        key={uniqueId()}
                        className={`${styles.charactersListItem}`}
                      >
                        <Link
                          to={`${ROUTES.CHARACTER.PATH}/${item.id}`}
                          className={styles.charactersListLink}
                        >
                          <CharacterCard info={item} />
                        </Link>
                      </li>
                    ))}
                </ul>
              </Col>
            </Row>
          )}
        </Col>
      </Row>
    </div>
  );
}
