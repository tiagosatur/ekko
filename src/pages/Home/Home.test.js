/* eslint-env jest */
import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import { renderWithProviders as render, testId } from "../../utils";

import Home from "./Home";

describe("Home component", () => {
  beforeEach(() => {
    cleanup();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<Home />);
    expect(asFragment(<Home />)).toMatchSnapshot();
  });

  // it("renders", () => {
  //   const { queryByTestId } = render(<Home />);

  //   expect(queryByTestId(testId.home.homepage)) ;
  // });
});
