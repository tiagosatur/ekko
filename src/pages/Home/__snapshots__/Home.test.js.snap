// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="homepage"
  >
    <div
      class="row"
    >
      <div
        class="col-xs-12"
      >
        <div
          class="container"
        >
          <div
            class="spinner"
          >
            <div />
            <div />
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
